lib/HTFeed/StorageAudit.pm

```diff
@@ -0,0 +1,103 @@
+package HTFeed::StorageAudit;
+
+use warnings;
+use strict;
+use Carp;
+use HTFeed::DBTools qw(get_dbh);
+use Log::Log4perl qw(get_logger);
+
+sub new {
+  my $class = shift;
+
+  my $object = {
+      @_,
+  };
+  if ( $class ne __PACKAGE__ ) {
+      croak "use __PACKAGE__ constructor to create $class object";
+  }
+  # check parameters
+  croak "invalid args" unless ($object->{bucket} and $object->{awscli});
+  bless( $object, $class );
+  return $object;
+}
+
+sub set_error {
+  my $self  = shift;
+  my $error = shift;
+
+  # log error w/ l4p
+  my $logger = get_logger( ref($self) );
+  $logger->error(
+    $error,
+    @_
+  );
+}
+
+# Iterates through feed_backups produces an error for each zip/xml not in AWS.
+# If limit is defined, checks that number of volumes.
+# Returns number of errors.
+sub run_not_in_aws_check {
+  my $self  = shift;
+  my $limit = shift;
+
+  my $err_count = 0;
+  my $s3 = HTFeed::Storage::S3->new(bucket => $self->{bucket},
+                                    awscli => $self->{awscli});
+  my $sql = 'SELECT namespace,id,version FROM feed_backups'.
+            ' WHERE path LIKE "s3://%"'.
+            ' AND deleted IS NULL'.
+            ' ORDER BY lastchecked ASC';
+  $sql .= "LIMIT $limit" if defined $limit;
+  foreach my $row (@{get_dbh()->selectall_arrayref($sql)}) {
+    my ($namespace, $id, $version) = @$row;
+    unless ($s3->s3_has("$namespace.$id.$version.zip")) {
+      $self->set_error('MissingFile', detail => "$namespace.$id.$version.zip not found in AWS");
```

Comment:
We probably should log this (and other errors) to the `feed_audit_detail` table for reporting (in addition to the regular logging)

Reply:
Added a second commit with feed_audit_detail logging.